### PR TITLE
Separated pool for background moves

### PR DESCRIPTION
**Pull request by excitoon**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (up to few sentences, not needed for non-significant PRs):

Added separated pool for background moves.

Progress on #7660 : Background moves can use all the background threads pool, making merges impossible, and lead to "Too many parts"

### CODE REVIEW

File: dbms/src/Interpreters/Context.cpp [hunk 1]
```diff
@@ -1489,6 +1491,14 @@ BackgroundProcessingPool & Context::getBackgroundPool()
     return *shared->background_pool;
 }
 
+BackgroundProcessingPool & Context::getMovePool()
+{
+    auto lock = getLock();
+    if (!shared->move_pool)
+        shared->move_pool.emplace(settings.move_pool_size);
```
Comment (filimonov) [inline dbms/src/Interpreters/Context.cpp]: maybe names of threads & log channel inside BackgroundProcessingPool should be parametrized to distinguish moving & normal background pools?

File: dbms/src/Storages/StorageMergeTree.cpp [hunk 2]
```diff
@@ -99,7 +99,7 @@ void StorageMergeTree::startup()
     /// NOTE background task will also do the above cleanups periodically.
     time_after_previous_cleanup.restart();
     merging_mutating_task_handle = global_context.getBackgroundPool().addTask([this] { return mergeMutateTask(); });
-    moving_task_handle = global_context.getBackgroundPool().addTask([this] { return movePartsTask(); });
+    moving_task_handle = global_context.getMovePool().addTask([this] { return movePartsTask(); });
```
Comment (filimonov) [inline dbms/src/Storages/StorageMergeTree.cpp]: Not sure if we should start 8 extra threads even for cases when no multidisk config is in place.

File: dbms/src/Core/Settings.h [hunk 3]
```diff
@@ -73,6 +73,7 @@ struct Settings : public SettingsCollection<Settings>
     M(SettingBool, use_uncompressed_cache, true, "Whether to use the cache of uncompressed blocks.") \
     M(SettingBool, replace_running_query, false, "Whether the running request should be canceled with the same id as the new one.") \
     M(SettingUInt64, background_pool_size, 16, "Number of threads performing background work for tables (for example, merging in merge tree). Only has meaning at server startup.") \
+    M(SettingUInt64, move_pool_size, 8, "Number of threads performing background moves for tables. Only has meaning at server startup.") \
```
Comment (filimonov) [inline dbms/src/Core/Settings.h]: maybe also `background_...`?

Comment (excitoon) [reply dbms/src/Core/Settings.h]: +

File: dbms/src/Storages/StorageMergeTree.cpp (hunk 2 above)
Comment (excitoon) [reply dbms/src/Storages/StorageMergeTree.cpp]: +

File: dbms/src/Interpreters/Context.cpp (hunk 1 above)
Comment (excitoon) [reply dbms/src/Interpreters/Context.cpp]: +

**Comment by filimonov** [comment]
what was the reason of data race?

File: dbms/src/Storages/MergeTree/BackgroundProcessingPool.cpp [hunk 4]
```diff
@@ -61,9 +61,12 @@ void BackgroundProcessingPoolTaskInfo::wake()
 }
 
 
-BackgroundProcessingPool::BackgroundProcessingPool(int size_) : size(size_)
+BackgroundProcessingPool::BackgroundProcessingPool(int size_, const char * thread_name_)
+    : size(size_)
+    , thread_name(thread_name_)
 {
-    LOG_INFO(&Logger::get("BackgroundProcessingPool"), "Create BackgroundProcessingPool with " << size << " threads");
+    logger = &Logger::get(String("BackgroundProcessingPool[") + thread_name + "]");
```
Comment (filimonov) [inline dbms/src/Storages/MergeTree/BackgroundProcessingPool.cpp]: not sure about braces, may be bit safer to use alphanumerical (log parsers, thread name in /proc/ etc.)

File: dbms/src/Interpreters/Context.cpp [hunk 5]
```diff
@@ -1485,10 +1487,18 @@ BackgroundProcessingPool & Context::getBackgroundPool()
 {
     auto lock = getLock();
     if (!shared->background_pool)
-        shared->background_pool.emplace(settings.background_pool_size);
+        shared->background_pool.emplace(settings.background_pool_size, "BackgrProcPool");
```
Comment (filimonov) [inline dbms/src/Interpreters/Context.cpp]: Parameter looks like uneeded.

**Review by filimonov — APPROVED** [review]

**Review by KochetovNicolai — APPROVED** [review]

**Pull request merged**